src/test/java/com/redhat/devtools/lsp4ij/features/formatting/TypeScriptEditorImprovementsTest.java

```diff
@@ -545,4 +546,70 @@ export class Foo {
     private static String adjustIndent(@NotNull String fileBody, boolean useTabCharacter) {
         return useTabCharacter ? fileBody.replace("    ", "\t") : fileBody;
     }
+
+    // These tests exercise the behavior of LSPImprovedTextMateNestedBracesTypedHandler
+
+    public void testImprovedTextMateNestedBraces() {
+        String fileBody = """
+                function foo() {
+                    invokePromise()
+                        .then()
+                }
+                """;
+        int initialOffset = fileBody.indexOf(".then(") + ".then(".length();
+
+        myFixture.configureByText(TEST_FILE_NAME, fileBody);
+        initializeLanguageServer();
+
+        Editor editor = myFixture.getEditor();
+        Document document = editor.getDocument();
+
+        CaretModel caretModel = editor.getCaretModel();
+        caretModel.moveToOffset(initialOffset);
+
+        // Type four open parens and we should end up five balanced parens
+        for (int iteration = 1; iteration <= 4; iteration++) {
+            EditorTestUtil.performTypingAction(editor, '(');
+            assertEquals(fileBody.replace(".then()", ".then(" + StringUtil.repeatSymbol('(', iteration) + StringUtil.repeatSymbol(')', iteration) + ")"), document.getText());
+            assertEquals(initialOffset + iteration, caretModel.getOffset());
+        }
+
+        // Type five close parens and nothing should be inserted but the caret should be advanced each time
+        String fiveBalancedParens = fileBody.replace(".then()", ".then" + StringUtil.repeatSymbol('(', 5) + StringUtil.repeatSymbol(')', 5));
+        initialOffset = caretModel.getOffset();
+        for (int iteration = 1; iteration <= 5; iteration++) {
+            EditorTestUtil.performTypingAction(editor, ')');
+            assertEquals(fiveBalancedParens, document.getText());
+            assertEquals(initialOffset + iteration, caretModel.getOffset());
+        }
+    }
+
+    public void testImprovedTextMateNestedBracesDisabled() {
```

Comment:
Confirms that the new TextMate-specific nested braces improvements can be disabled via client config for user-defined language servers.